### PR TITLE
nixos/frp: init frp module with a test

### PR DESCRIPTION
**Pull request by zaldnoay**
## Description of changes

Supersedes #235419.

Adding a [frp](https://github.com/fatedier/frp) module.
The original author doesn't seem to be responding to the reviewer's comments, so I'll take over updating that PR.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/frp-client.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp-client.nix:60]: `"AF_INET AF_INET6"` may generate a working service file, but it is wrong from the perspective of nix lang.  I believe its a typoe.

**Review by jian-lin — CHANGES_REQUESTED** [review]
Please check my reviews in the original PR about:
- dependencies on the network(-online).target
- redundant hardening options implied by `DynamicUser` (On, why do you remove `DynamicUser`?  Please add it back.  It should be used if possible.)


Suggestions:
- `services.frp-client` and `services.frp-server` are almost the same.  It would be great to merge them into one module.
- Implement secret-handling logic

**Comment by zaldnoay** [comment]
> dependencies on the network(-online).target

The original author marked this issue as solved, so I missed this comment. I'll correct it later.


> * redundant hardening options implied by `DynamicUser` (On, why do you remove `DynamicUser`?  Please add it back.  It should be used if possible.)

Oh, I misunderstood what you meant in the original PR, and I will modify the hardening options later.

> * `services.frp-client` and `services.frp-server` are almost the same.  It would be great to merge them into one module.

Good idea, but if I refactor the frp module, do I need to create a new PR or update it to this PR?

File: nixos/modules/services/networking/frp-client.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp-client.nix:60]: I believe this is a typo by the original author and will be corrected later.

**Comment by jian-lin** [comment]
> if I refactor the frp module, do I need to create a new PR or update it to this PR

I do not see the need for a new PR.

File: nixos/modules/services/networking/frp-client.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp-client.nix:38]: `man systemd.special`:

> Units that strictly require a configured network connection should pull in network-online.target (via a Wants= type dependency) and order themselves after it.

```suggestion
        wants = [ "network-online.target" ];
        after = [ "network-online.target" ];
```

**Review by jian-lin — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/frp-client.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp-client.nix:38]: Thank you. I've updated it. Later I will rework this PR to combine two frp modules.

**Comment by Janik-Haag** [comment]
>  The original author doesn't seem to be responding to the reviewer's comments, so I'll take over updating that PR.

Thank you for doing this, to be honest I kinda forgot about it :see_no_evil:

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (jian-lin) [inline nixos/doc/manual/release-notes/rl-2311.section.md:19]:
```suggestion
- [frp](https://github.com/fatedier/frp), a fast reverse proxy to help you expose a local server behind a NAT or firewall to the Internet. Available as [services.frp](#opt-services.frp.enable).
```

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:15]: Something is missing in `""`.

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:21]: I think we should force users to set `role` by not specifying `default`.

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:58]: All variables except `serviceCapability` are only used once so there is no reason to bind them there.

`lib` should be removed for consistency sake.

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:66]: See the manual of `Description` in `man systemd.unit`.

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:75]: `DynamicUser` is not usually considered as an hardening option.
```suggestion
            DynamicUser = true;
            # Hardening
            UMask = serviceUMask;
```

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:98]: @Janik-Haag Do you want to be a maintainer?

**Review by jian-lin — CHANGES_REQUESTED** [review]
Generally looks good.

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Done.

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Default have been remove.

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Done

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Already changed it. See if you have another opinion.

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:15]:
```suggestion
      enable = mkEnableOption (lib.mdDoc "frp.");
```

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: These binds have been move to service unit. But I would like to keep `executableFile` here to to make the `ExecStart` more clearer. What do you think?

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:51]:
```suggestion
      serviceCapability = optionals isServer [ "CAP_NET_BIND_SERVICE" ];
```

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:60]:
```suggestion
          description = "A fast reverse proxy frp ${cfg.role}";
```

**Review by jian-lin — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Thank you. I forgot this.

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Done

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix]: Done

**Review by jian-lin — COMMENTED** [review]
- link the test to the package https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package
- rebase your commits into one or two: one for adding the module and another for adding the test

File: nixos/modules/services/networking/frp.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/frp.nix:92]: Thank you for asking but since I didn't use the tool in the last few month I would probably say no.

**Comment by zaldnoay** [comment]
> * link the test to the package https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package
> 
> * rebase your commits into one or two: one for adding the module and another for adding the test

Both are done. Please check it again.

File: nixos/modules/services/networking/frp.nix
Comment (jian-lin) [inline nixos/modules/services/networking/frp.nix:69]: According to the output of `systemd-analyze security frp`, it seems that more hardening options can be added.  @zaldnoay do you want to work on this?

However, I think the current version is good enough to be merged since most of existing NixOS modules are not hardened at all.

**Review by jian-lin — APPROVED** [review]

File: nixos/modules/services/networking/frp.nix
Comment (zaldnoay) [reply nixos/modules/services/networking/frp.nix:69]: This option is copied from the previous PR, I think setting it to `0077` should work. However, this PR has been merged, so I would not likt to modify it for now.

**Pull request merged**
